Source/KSCrash/Source/KSCrash/Recording/Tools/BSG_KSMachHeaders.m

```diff
@@ -11,6 +11,118 @@
 #import <Foundation/Foundation.h>
 #import "BSG_KSDynamicLinker.h"
 #import "BSG_KSMachHeaders.h"
+#import "BugsnagPlatformConditional.h"
+
+// MARK: - Locking
+
+// Pragma's hide unavoidable (and expected) deprecation/unavailable warnings
+_Pragma("clang diagnostic push")
+_Pragma("clang diagnostic ignored \"-Wunguarded-availability\"")
+static os_unfair_lock bsg_mach_binary_images_access_lock_unfair = OS_UNFAIR_LOCK_INIT;
+_Pragma("clang diagnostic pop")
+
+_Pragma("clang diagnostic push")
+_Pragma("clang diagnostic ignored \"-Wdeprecated-declarations\"")
+static OSSpinLock bsg_mach_binary_images_access_lock_spin = OS_SPINLOCK_INIT;
+_Pragma("clang diagnostic pop")
+
+// A static copy of the OS for comparison efficiency
+static NSOperatingSystemVersion bsg_os_version;
+
+// Lock helpers.  These use bulky Pragmas to hide warnings so are in their own functions for clarity.
+
+void bsg_spin_lock() {
+    _Pragma("clang diagnostic push")
+    _Pragma("clang diagnostic ignored \"-Wdeprecated-declarations\"")
+    OSSpinLockLock(&bsg_mach_binary_images_access_lock_spin);
+    _Pragma("clang diagnostic pop")
+}
+
+void bsg_spin_unlock() {
+    _Pragma("clang diagnostic push")
+    _Pragma("clang diagnostic ignored \"-Wdeprecated-declarations\"")
+    OSSpinLockUnlock(&bsg_mach_binary_images_access_lock_spin);
+    _Pragma("clang diagnostic pop")
+}
+
+void bsg_unfair_lock() {
+    _Pragma("clang diagnostic push")
+    _Pragma("clang diagnostic ignored \"-Wunguarded-availability\"")
+    os_unfair_lock_lock(&bsg_mach_binary_images_access_lock_unfair);
+    _Pragma("clang diagnostic pop")
+}
+
+void bsg_unfair_unlock() {
+    _Pragma("clang diagnostic push")
+    _Pragma("clang diagnostic ignored \"-Wunguarded-availability\"")
+    os_unfair_lock_unlock(&bsg_mach_binary_images_access_lock_unfair);
+    _Pragma("clang diagnostic pop")
+}
+
+// Lock and unlock sections of code
+
+void bsg_dyld_cache_lock() {
```

Comment:
This function is a candidate for the use of [isOperatingSystemAtLeastVersion](https://developer.apple.com/documentation/foundation/nsprocessinfo/1414876-isoperatingsystematleastversion?language=objc), which would simplify the code and minimise the number of lock calls/return statements:

```objc
- (BOOL)spinLockSupported {
    NSOperatingSystemVersion minSdk;

    #if BSG_PLATFORM_IOS
        minSdk.majorVersion = 10;
        minSdk.minorVersion = 12;
        minSdk.minorVersion = 0;
    #elif
        // ... handle other platforms
    #endif
    return [NSProcessInfo isOperatingSystemAtLeastVersion:minSdk];
}

void bsg_dyld_cache_lock() {
    if (spinLockSupported) {
        bsg_spin_lock();
    } else {
        bsg_unfair_lock();
    }
}
```